Performance improvement - update at most at the "Update rate"

wxmerkt opened the pull request:
This PR achieves two things:

1. It only renders when triggered by the RViz `update` callback: This way, we respect the "Update rate" setting set in the Global Options in RViz. Currently it renders on every incoming message - which for the Anymal controller is 400 Hz - and at ~1.5-2ms per update (due to the pointer resetting and PinocchioLinkUpdater callbacks), this quickly blocks the GPU and results in dropping RViz fps.
2. It catches a few invalid user inputs in the URDF and arrow settings which previously led to RViz crashing

cmastalli commented:
Thanks for the improvement :)

wxmerkt commented on include/whole_body_state_rviz_plugin/WholeBodyStateDisplay.h:
```diff
@@ -205,7 +212,7 @@ private Q_SLOTS:
 
   enum CoMStyle { REAL, PROJECTED }; //!< CoM visualization style
-  bool com_real_;     //!< Label to indicates the type of CoM display (real or
+  bool com_real_; //!< Label to indicates the type of CoM display (real or
```
Thanks! I've just added another commit with the formatting

wxmerkt commented:
This should be ready to merge as well :-)

cmastalli commented on include/whole_body_state_rviz_plugin/WholeBodyStateDisplay.h:
```diff
@@ -1,14 +1,18 @@
 #ifndef WHOLE_BODY_STATE_RVIZ_PLUGIN_WHOLE_BODY_STATE_DISPLAY_H
 #define WHOLE_BODY_STATE_RVIZ_PLUGIN_WHOLE_BODY_STATE_DISPLAY_H
 
+#include "whole_body_state_rviz_plugin/ArrowVisual.h"
```
Do we need to move these header before pinocchio stuffs?

I often follow include the internal headers after the global ones.

wxmerkt replied:
The formatting did the re-order as far as I can recall

cmastalli commented on src/WholeBodyStateDisplay.cpp:
```diff
@@ -279,6 +290,7 @@ void WholeBodyStateDisplay::loadRobotModel() {
   if (content.empty()) {
     clearRobotModel();
     setStatus(StatusProperty::Error, "URDF", "URDF is empty");
+    ROS_ERROR("URDF is empty");
```
Why do we need to use ROS_ERROR?

Note that we use `setStatus` which should print the error message in Rviz

wxmerkt replied:
It was just easier to see during debugging than opening the status when it initially segfaulted. Happy to remove

cmastalli replied:
Ok, I see, it is fine to keep it :)

cmastalli commented on src/WholeBodyStateDisplay.cpp:
```diff
@@ -289,12 +301,22 @@ void WholeBodyStateDisplay::loadRobotModel() {
   if (!descr.initString(robot_model_)) {
     clearRobotModel();
     setStatus(StatusProperty::Error, "URDF", "Failed to parse URDF model");
+    ROS_ERROR("Failed to parse URDF model");
```
same question here (see above)

cmastalli commented on src/WholeBodyStateDisplay.cpp:
```diff
@@ -289,12 +301,22 @@ void WholeBodyStateDisplay::loadRobotModel() {
+    error_msg += e.what();
+    setStatus(StatusProperty::Error, "Pinocchio-URDFParser",
+              QString::fromStdString(error_msg));
+    ROS_ERROR_STREAM(error_msg);
```
and same question here

wxmerkt replied:
This error was too long to be read properly otherwise I believe

cmastalli reviewed (commented):
I have a few questions about this PR.
Additionally, do you know why it is not passing the CI?

wxmerkt commented:
I fixed the build issues and changed the CI. It will be live after this PR has been merged. Example output: https://github.com/wxmerkt/state_rviz_plugin/runs/2200889085?check_suite_focus=true

cmastalli reviewed (approved):
Let's merge it if you've finished with it.

wxmerkt commented:
I'm finished with it :-)

cmastalli commented:
@wxmerkt -- thanks for improving the plugin.
